packages/block-editor/src/hooks/block-style-variation.js

```diff
@@ -0,0 +1,156 @@
+import { useSelect } from '@wordpress/data';
+import { useContext, useMemo } from '@wordpress/element';
+
+/**
+ * Internal dependencies
+ */
+import {
+	GlobalStylesContext,
+	toStyles,
+	getBlockSelectors,
+} from '../components/global-styles';
+import { useStyleOverride } from './utils';
+import { store as blockEditorStore } from '../store';
+import { globalStylesDataKey } from '../store/private-keys';
+
+/**
+ * Get the first block style variation that has been registered from the class string.
+ *
+ * @param {string} className        CSS class string for a block.
+ * @param {Array}  registeredStyles Currently registered block styles.
+ *
+ * @return {string|null} The name of the first registered variation.
+ */
+function getVariationNameFromClass( className, registeredStyles = [] ) {
+	// The global flag affects how capturing groups work in JS. So the regex
+	// below will only return full CSS classes not just the variation name.
+	const matches = className?.match( /\bis-style-(?!default)(\S+)\b/g );
+
+	if ( ! matches ) {
+		return null;
+	}
+
+	for ( const variationClass of matches ) {
+		const variation = variationClass.substring( 9 ); // Remove 'is-style-' prefix.
+		if ( registeredStyles.some( ( style ) => style.name === variation ) ) {
+			return variation;
+		}
+	}
+	return null;
+}
+
+function useBlockSyleVariation( name, variation, clientId ) {
+	// Prefer global styles data in GlobalStylesContext, which are available
+	// if in the site editor. Otherwise fall back to whatever is in the
+	// editor settings and available in the post editor.
+	const { merged: mergedConfig } = useContext( GlobalStylesContext );
+	const { globalSettings, globalStyles } = useSelect( ( select ) => {
+		const settings = select( blockEditorStore ).getSettings();
+		return {
+			globalSettings: settings.__experimentalFeatures,
+			globalStyles: settings[ globalStylesDataKey ],
+		};
+	}, [] );
+
+	return useMemo( () => {
+		const styles = mergedConfig?.styles ?? globalStyles;
+		const variationStyles =
+			styles?.blocks?.[ name ]?.variations?.[ variation ];
+
+		return {
+			settings: mergedConfig?.settings ?? globalSettings,
+			// The variation style data is all that is needed to generate
+			// the styles for the current application to a block. The variation
+			// name is updated to match the instance specific class name.
+			styles: {
+				blocks: {
+					[ name ]: {
+						variations: {
+							[ `${ variation }-${ clientId }` ]: variationStyles,
+						},
+					},
+				},
+			},
+		};
+	}, [
+		mergedConfig,
+		globalSettings,
+		globalStyles,
+		variation,
+		clientId,
+		name,
+	] );
+}
+
+// Rather than leveraging `useInstanceId` here, the `clientId` is used.
+// This is so that the variation style override's ID is predictable
+// when the order of applied style variations changes.
+function useBlockProps( { name, className, clientId } ) {
+	const { getBlockStyles } = useSelect( blocksStore );
+
+	const registeredStyles = getBlockStyles( name );
```

Comment:
Calling a store selector during render could lead to some unexpected results. 

The hook isn't subscribed to block style updates, and it will only fetch new values when the consumer or `useBlockSyleVariation` triggers a re-render.

Reply:
Would there be a risk when [new styles are registered/deregistered](https://github.com/WordPress/gutenberg/blob/8f74d5c297076f2a247b452f455a9bbb72bc055e/packages/blocks/src/store/reducer.js#L203)? 

I don't know yet if it's worth it, but I think by refactoring `getBlockSelectors` or creating a new function that accepts a single block type and `registeredStyles` (and not the function), it'd be possible to call `getBlockStyles` in the store subscription callback, e.g., 


```js
const blockType = useMemo(  
    () =>  
       getBlockTypes().find(  
          ( { name: blockName } ) => blockName === name  
       ),  
    [ name ]  
);  
  
const { registeredStyles } = useSelect(  
    ( select ) => {  
       const { getBlockStyles } = select( blocksStore );  
       return {  
          registeredStyles: getBlockStyles( name ),  
       };  
    },  
    [ name ]  
);

...


const blockSelectors = getBlockSelectorForBlock(  
    blockType,  
    registeredStyles,  
    clientId  
);
```